BlueDBOnDisk/src/main/java/org/bluedb/disk/collection/CollectionEntityIterator.java

```diff
@@ -40,30 +43,42 @@ private long calculateEndGroupingValueOfCompletedSegments(Range range, boolean b
 	}
 
 	@Override
-	public void close() {
-		if (segmentIterator != null) {
+	public synchronized void close() {
+		if(!hasClosed.getAndSet(true) && segmentIterator != null) {
 			segmentIterator.close();
 		}
 	}
 
 	@Override
-	public boolean hasNext() {
+	public synchronized boolean hasNext() {
+		if (hasClosed.get()) {
+			throw new RuntimeException("CollectionEntityIterator has already been closed");
+		}
+		
 		if (next == null) {
 			next = nextFromSegment();
 		}
 		return next != null;
 	}
 
 	@Override
-	public BlueEntity<T> peek() {
+	public synchronized BlueEntity<T> peek() {
+		if (hasClosed.get()) {
+			throw new RuntimeException("CollectionEntityIterator has already been closed");

```

Comment:
That is what `CollectionValueIterator` does already so I just followed the pattern. It contains the two classes that you mention and since it doesn't allow its methods to be called it should shield these methods from being called. But I thought it was worth a little extra protection since we definitely want to avoid getting a lock if we're closed since close will not be called again. I'd be find to change it to some other type of exception if you'd like.